mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp

```diff
@@ -1740,3 +1741,294 @@ Value mlir::rock::addPassThroughIndices(OpBuilder &b, Value transformed,
+  auto dims = Type == DimType::Upper ? tr.getUpperDims() : tr.getLowerDims();
+  assert(names.size() == dims.size());
+  for (auto [idx, dim] : llvm::enumerate(dims)) {
+    if (llvm::is_contained(preservedDims, dim)) {
+      preseredNames.push_back(names[idx]);
+    }
+  }
+}
+
+template <DimType Type>
+SmallVector<uint32_t> getDifference(rock::TransformAttr tr,
+                                    TransformAttrArgs &args) {
+  SmallVector<uint32_t> difference;
+  auto dims = Type == DimType::Upper ? tr.getUpperDims() : tr.getLowerDims();
+  const auto &preserved = std::get<Type>(args.preservedDims);
+  for (auto dim : dims) {
+    if (!llvm::is_contained(preserved, dim)) {
+      difference.push_back(dim);
+    }
+  }
+  return difference;
+}
+
+template <DimType Type>
+void remapDims(
+    std::vector<TransformAttrArgs> &argsVector,
+    std::pair<SmallVector<uint32_t>, SmallVector<uint32_t>> &removedDims) {
+  auto &sortedDeletedDims = std::get<Type>(removedDims);
+  llvm::sort(sortedDeletedDims);
+  for (auto &args : argsVector) {
+    auto &preserved = std::get<Type>(args.preservedDims);
+    for (auto [idx, dim] : llvm::enumerate(preserved)) {
+      size_t leastUpperBoundIdx = 0;
+      while ((leastUpperBoundIdx < sortedDeletedDims.size()) &&
+             (dim > sortedDeletedDims[leastUpperBoundIdx])) {
+        ++leastUpperBoundIdx;
+      }
+      preserved[idx] -= leastUpperBoundIdx;
+    }
+  }
+}
+
+FailureOr<rock::TransformMapAttr>
+removeUpperDimsFromMap(OpBuilder &b, rock::TransformMapAttr trMap,
+                       SetVector<int64_t> &removeIndicesSet,
+                       llvm::SmallVector<int64_t> &origUpperBounds,
+                       llvm::SmallVector<int64_t> &origLowerBounds) {
+
+  origLowerBounds =
+      llvm::SmallVector<int64_t>(trMap.getLowerBounds().asArrayRef());
+
+  SetVector<int64_t> newRemoveIndicesSet;
+  SmallVector<TransformAttr> newOps;
+  std::pair<SmallVector<uint32_t>, SmallVector<uint32_t>> removedDims = {};
+
+  std::vector<TransformAttrArgs> argsVector;
+  for (auto tr : trMap.getOps()) {
+    TransformAttrArgs args;
+    args.type = tr.getType();
+    SmallVector<uint32_t> &preservedUpperDims =
+        std::get<DimType::Upper>(args.preservedDims);
+    SmallVector<uint32_t> &preservedLowerDims =
+        std::get<DimType::Lower>(args.preservedDims);
+
+    preservedUpperDims =
+        getPreservedIndices<DimType::Upper>(tr, removeIndicesSet);
+
+    const bool mustBePreserved =
+        preservedUpperDims.size() == tr.getUpperDims().size();
+    const bool mustBeCompletelyRemoved = preservedUpperDims.empty();
+    const bool mustBeModified = !mustBePreserved && !mustBeCompletelyRemoved;
+
+    // compute which lower dimensions must be preserved
+    switch (args.type) {
+    case TransformType::AddDim:
+    case TransformType::ConstDim:
+    case TransformType::Broadcast:
+    case TransformType::Merge: {
+      assert(!mustBeModified && "assume always one-to-many relation");
```

Comment:
Hi @manupak,
This is the assumption within a single `TransformMapAttr`. It means that: `<newDim> -> <dimX, dimY, dimZ>, where `newDim` is a new upper dimension;  `dimX`, `dimY`, `dimZ` - some lower dimensions.

Reply:
Well, `Merge` is a one-to-(N >= 1) operation.

I think that's the assert not being phrased right.

Reply:
> Well, Merge is a one-to-(N >= 1) operation.

Yes. The assertion says that the transform must be either preserved or completely removed. So, there is no possible scenario to `modify` merge. I would suggest to just remove the string literal in the `assert`. I hope it will remove this ambiguity

Reply:
Changed to `must be preserved or removed completely`